File: keyboards/windows/LSDKeyboard/LSDKeyboard/PairCollector.cs
using Microsoft.Data.Sqlite;

namespace LSDKeyboard;

// Schema is identical to iOS / macOS / Android so all platform databases
// can be merged into one corpus without any transformation.
//
// Database location: %APPDATA%\LSDKeyboard\lsd_pairs.sqlite

public sealed class PairCollector : IDisposable
{
    public static readonly PairCollector Instance = new();

    private readonly SqliteConnection _db;

    private PairCollector()
    {
        var dir = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            "LSDKeyboard");
        Directory.CreateDirectory(dir);

        _db = new SqliteConnection($"Data Source={Path.Combine(dir, "lsd_pairs.sqlite")}");
        _db.Open();
        CreateSchema();
    }

    // -------------------------------------------------------------------------
    // Schema

    private void CreateSchema()
    {
        Exec("""
            CREATE TABLE IF NOT EXISTS pairs (
                id           INTEGER PRIMARY KEY AUTOINCREMENT,
                lsd_input    TEXT    NOT NULL,
                roman_output TEXT    NOT NULL,
                source       TEXT    NOT NULL DEFAULT 'correction',
                created_at   REAL    NOT NULL,
                contributed  INTEGER NOT NULL DEFAULT 0
            );
            CREATE INDEX IF NOT EXISTS idx_contributed ON pairs (contributed);
            """);
    }

    // -------------------------------------------------------------------------
    // Recording

    public void RecordAccepted(string lsd, string roman) =>
        Insert(lsd, roman, "accepted");

    public void RecordCorrection(string lsd, string correctedRoman) =>
        Insert(lsd, correctedRoman, "correction");

    // Double-press substitution (e.g. سس → ے).
    // lsd_input    = secondary (what was inserted)
    // roman_output = primary key that triggered the rule
    public void RecordDoublePress(string primary, string secondary) =>
        Insert(secondary, primary, "double_press_windows");

    // -------------------------------------------------------------------------
    // Queries

    public int PendingCount()
    {
        using var cmd = Cmd("SELECT COUNT(*) FROM pairs WHERE contributed = 0");
        return Convert.ToInt32(cmd.ExecuteScalar());
    }

    public int TotalCount()
    {
        using var cmd = Cmd("SELECT COUNT(*) FROM pairs");
        return Convert.ToInt32(cmd.ExecuteScalar());
    }

    public IReadOnlyList<(long Id, string Lsd, string Roman)> FetchPending(int limit = 2000)
    {
        using var cmd = Cmd(
            "SELECT id, lsd_input, roman_output FROM pairs WHERE contributed = 0 LIMIT @lim");
        cmd.Parameters.AddWithValue("@lim", limit);

        var results = new List<(long, string, string)>();
        using var reader = cmd.ExecuteReader();
        while (reader.Read())
            results.Add((reader.GetInt64(0), reader.GetString(1), reader.GetString(2)));
        return results;
    }

    public void MarkContributed(IEnumerable<long> ids)
    {
        var list = ids.ToList();
        if (list.Count == 0) return;
        var placeholders = string.Join(",", list.Select((_, i) => $"@p{i}"));
        using var cmd = Cmd($"UPDATE pairs SET contributed = 1 WHERE id IN ({placeholders})");
        for (int i = 0; i < list.Count; i++)
            cmd.Parameters.AddWithValue($"@p{i}", list[i]);
        cmd.ExecuteNonQuery();
    }

    public void DeleteAll() => Exec("DELETE FROM pairs");

    // -------------------------------------------------------------------------
    // Helpers

    private void Insert(string lsd, string roman, string source)
    {
        using var cmd = Cmd(
            "INSERT INTO pairs (lsd_input, roman_output, source, created_at) VALUES (@l,@r,@s,@t)");
        cmd.Parameters.AddWithValue("@l", lsd);
        cmd.Parameters.AddWithValue("@r", roman);
        cmd.Parameters.AddWithValue("@s", source);
        cmd.Parameters.AddWithValue("@t", DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0);
        cmd.ExecuteNonQuery();
    }

    private SqliteCommand Cmd(string sql)
    {
        var cmd = _db.CreateCommand();
        cmd.CommandText = sql;
        return cmd;
    }

    private void Exec(string sql) { using var cmd = Cmd(sql); cmd.ExecuteNonQuery(); }

    public void Dispose() => _db.Dispose();
}

File: keyboards/android/LSDKeyboard/app/src/main/java/com/exordiumnetworks/lsdkeyboard/PairCollector.kt
package com.exordiumnetworks.lsdkeyboard

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

// Schema is identical to iOS/macOS so data from all platforms can be merged.
// DB is stored in the app's private data directory (no external storage permission needed).

class PairCollector private constructor(context: Context) :
    SQLiteOpenHelper(context.applicationContext, "lsd_pairs.sqlite", null, 1) {

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL("""
            CREATE TABLE IF NOT EXISTS pairs (
                id           INTEGER PRIMARY KEY AUTOINCREMENT,
                lsd_input    TEXT    NOT NULL,
                roman_output TEXT    NOT NULL,
                source       TEXT    NOT NULL DEFAULT 'correction',
                created_at   REAL    NOT NULL,
                contributed  INTEGER NOT NULL DEFAULT 0
            )
        """.trimIndent())
        db.execSQL("CREATE INDEX IF NOT EXISTS idx_contributed ON pairs (contributed)")
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {}

    // MARK: - Recording

    fun recordAccepted(lsd: String, roman: String) = insert(lsd, roman, "accepted")

    fun recordCorrection(lsd: String, @Suppress("UNUSED_PARAMETER") suggestedRoman: String, correctedRoman: String) {
        insert(lsd, correctedRoman, "correction")
    }

    // Double-press substitution: e.g. user pressed "س" twice → "ے" was inserted.
    // lsd_input  = secondary (what was actually inserted)
    // roman_output = primary key that was pressed (records which rule fired)
    fun recordDoublePress(primary: String, secondary: String) =
        insert(secondary, primary, "double_press_android")

    // MARK: - Queries

    fun pendingCount(): Int {
        return readableDatabase.rawQuery(
            "SELECT COUNT(*) FROM pairs WHERE contributed = 0", null
        ).use { c -> if (c.moveToFirst()) c.getInt(0) else 0 }
    }

    fun totalCount(): Int {
        return readableDatabase.rawQuery("SELECT COUNT(*) FROM pairs", null)
            .use { c -> if (c.moveToFirst()) c.getInt(0) else 0 }
    }

    fun fetchPending(limit: Int = 2000): List<Triple<Long, String, String>> {
        val results = mutableListOf<Triple<Long, String, String>>()
        readableDatabase.rawQuery(
            "SELECT id, lsd_input, roman_output FROM pairs WHERE contributed = 0 LIMIT ?",
            arrayOf(limit.toString())
        ).use { c ->
            while (c.moveToNext()) {
                results += Triple(c.getLong(0), c.getString(1), c.getString(2))
            }
        }
        return results
    }

    fun markContributed(ids: List<Long>) {
        if (ids.isEmpty()) return
        val placeholders = ids.joinToString(",") { "?" }
        writableDatabase.execSQL(
            "UPDATE pairs SET contributed = 1 WHERE id IN ($placeholders)",
            ids.map { it.toString() }.toTypedArray()
        )
    }

    fun deleteAll() = writableDatabase.execSQL("DELETE FROM pairs")

    // MARK: - Helpers

    private fun insert(lsd: String, roman: String, source: String) {
        writableDatabase.execSQL(
            "INSERT INTO pairs (lsd_input, roman_output, source, created_at) VALUES (?,?,?,?)",
            arrayOf<Any>(lsd, roman, source, System.currentTimeMillis() / 1000.0)
        )
    }

    companion object {
        @Volatile private var instance: PairCollector? = null

        fun getInstance(context: Context): PairCollector =
            instance ?: synchronized(this) {
                instance ?: PairCollector(context).also { instance = it }
            }
    }
}
